doc/man-source/pronto2lirc.1
.TH PRONTO2LIRC "1" "Last change: Oct 2016" "pronto2lirc @version@" "User Commands"
.SH NAME
pronto2lirc - convert pronto-encoded data to lircd.conf format
.SH SYNOPSIS
.B pronto2lirc <\fIinputfile\fR>
.SH DESCRIPTION
.P
Converts an input file to a lircd.conf formatted file. Inputfile can be:
.IP \- 2
An xml file output by CCFTools/CCFDecompiler.
.IP \- 2
Text file where each line contains the name of the button and
all codes associated with it:
.nf

         Button1:0000 00ac 000b 00de ...

.if
.P
\fIpronto2lirc\fR writes the result as a lircd.conf file  to the current
directory containing all the Pronto codes extracted from the input file.

.SH OPTIONS
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.SH "SEE ALSO"
.P
lircd.conf(5)


NEWS
0.9.4c 22/10/16
 * New curl_poll files adds a new MIT license. Lirc now carries
   GPL, BSD and MIT licenses.
 * Several patches to make lirc work again on macos (#226, #227).
 * New tool irtext2udp generates udp plugin data (#197).
 * Fix FTBS errors on linux kernel >= 4.8.0 (#232).
 * plugins: girs: Add udp/receive support (#201).
 * mode2: Fix broken --raw option (#223).
 * lirc-make-devinput: Fix that two remotes has the same name (#222).
 * lircd: Warn for duplicate remote definitions (#222).
 * lircd: Retry temporary unavailable write sockets (#221).
 * contrib: Fix usb devices' acl permissions (rhbz #1364744).
 * lirc-setup: Fix crasher bugfix (Fedora abort trace).
 * contrib: disable udev /dev/uinput rule requiring lirc user.
 * plugins: audio_alsa: Fix byte truncating in 16-bit data (#218):
 * plugins: irtoy: improve error handling (#220).
 * Minor documentation fixes.

0.9.4b 9/8/16
  * lircd-uinput: A multitude of fixes (#213, #211, #177). Adds new options.
  * lirc_client:  Fix "Can't redirect irsend output" (#207).
  * lirc_client: read_string() not terminating buffer contents (#216)
  * Remove a single python2 dependency [4e3c97c].
  * irrecord: Fix under-sized buffer causing crash (#208).
  * systemd: Add hardening options comments to services (#204).
  * lircmd: Options, udev examples and manpage update.A
  * lib: Add missing Makefile.am dep (#210).
  * doc: Manpages and configuration guide fixes.
  * build: Don't define docdir - autoconf defines docdir (#215).
  * build: support runstatedir configure option (#212).

0.9.4a 22/6/16
  * Fix syntax errors in systemd unit files (rhbz: 1350750).
  * irrecord: Enhance "Don't use devinput" message (#180).
  * plugins/uirt2_raw: Reset device to UIR mode on deinit.
  * plugins/girs.c: Enable reception after sending.
  * lirc.org: Updates and bugfixes
  * lircd, liblirc_client: Fix freeaddrinfo handling (#195).
  * Don't build commandir unless we have usb.h (#191).
  * lirc_options.conf: debug => loglevel (#199).
  * lib: use proper linking method to avoid parallel build issue
  * tools: make_rel_symlink.py can also use python2
  * plugins/devinput: Make the list-devices support available.
  * configure.ac: Bugfix (BSD setting reported wrong).

0.9.4 22/5/16
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4
  * Digest of 0.9.4pre1, 0.9.4pre2

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * The lirc.4 manpage (/dev/lirc* interface) is upstreamed to the
    linux kernel manpages project (#164). Old page lirc.4l still
    lingers in attic/ but is not installed.
  * Upstream selinux updated to match current lircd requirements.
    See: https://github.com/fedora-selinux/selinux-policy/pull/41

  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
    README.md in the debian branch (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * New transmit-only FTDI driver with enhanced accuracy and hardware
    support. See ftdix driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lirc-setup tool (#123, partly):
    - Add drvctl device enumeration support.
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * Several drivers including default and devinput has enhanced
    enumeration support, usable with mode2 -l and lirc-setup.
  * devinput.lircd.conf is generated at build time to match host kernel,
    new script lirc-make-devinput shipped to re-generate it (#153)
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).
  * Several new examples in contrib/, notably lircrc templates for
    many popular LIRC applications and improved udev rules.
  * The default driver does not require root perms if the protocol
    is correct (#133).
  * Add an "auto" device to devinput driver for single-device
    usecase (#134).
  * irrecord has been taught to handle remotes without gap (#98).
  * The --uinput lircd option is deprecated, obsoleted by the
    lircd.uinput.service systemd service. See lircd-uinput(8).
  * CLI options for mode2, xmode2 and irrecord has been streamlined.
    All tools falls back to lircd defaults where it makes sense.

  * The driver API has been expanded with functions to enumerate
    devices. See configs/README and lib/driver.h.
  * New logging API with hooks for enhanced messages and filtering
    (also #129).
  * The out-of-tree build interface is enhanced with hooks for external
    drivers to rebuild documentation and add a config file.

  * lircd: Fix too-short gap  when sending/blasting (#158).
  * lircd: Fix typo causing lost deinit_func() calls (#179).
  * irrecord fixes:
    - Fix broken namespace checks on recent kernels (#187).
    - Buggy handling of not decoded buttons in irrecord (#146).
    - Repeated buttons problems (#149)

tools/pronto2lirc.py
#
# A tool for converting Pronto format hex codes to lircd.conf format. Version 1.11
#
# Version History:
#       1.11 - Made more resiliant against whitespace imperfections in input files
#       1.1  - Added support for CCFTools/CCFDecompiler XML files and multiple devices
#       1.0  - Initial release
#
# Copyright by Olavi Akerman <olavi.akerman@gmail.com>
#
# pronto2lirc is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#

import string

class CodeSequence:           # Handles codesequences parsing and conversion

    def ProcessPreamble(self,sPreamble):
        if sPreamble[0]<>"0000":
            raise "Formats other than starting with 0000 are not supported!"

        self.dIRFrequency=1000000/(long(sPreamble[1],16) * 0.241246) # Frequency of the IR carrier in Khz

        self.lOnceSequenceLength=long(sPreamble[2],16)          # No of pulses that is sent once when button is pressed
        self.lRepeatableSequenceLength=long(sPreamble[3],16)    # No of pulses that are repeatable while button pressed

    def CreatePulses(self,sItems):
        self.dPulseWidths=[]             # Table of Pulse widths. Length is repsented in microseconds

        for i in sItems:
            self.dPulseWidths.append(1000000*long(i,16)/self.dIRFrequency) # Convert pulse widths to uS

        if len(self.dPulseWidths)<>2*(self.lOnceSequenceLength+self.lRepeatableSequenceLength):
            raise "Number of actual codes does not match the header information!"

    def AnalyzeCode(self,sCodeName,sHexCodes):

        # sHexTable=sHexCodes.split()
        s=string.join(sHexCodes.split(),'') # Remove whitespace formatting between sequences

        sHexTable=[]

        while s<>'':                        # Re-split into group of 4
            sHexTable.append(s[:4])
            s=s[4:]

        self.sCodeName=sCodeName.rstrip()  # Name of the Code associated with code sequence

        self.ProcessPreamble(sHexTable[:4]) # First four sequences make up Preamble
        self.CreatePulses(sHexTable[4:])    # The rest are OnceSequence + RepeatableSequence
        return self.dPulseWidths[-1]        # Final gap=last off signal length

    def WriteCodeSection(self,fOut):
        fOut.write('\n\t\t\tname '+self.sCodeName.strip().replace(' ','')+'\n')  # Can't contain whitespace
        for i in range(len(self.dPulseWidths)-1):   # Do not write the last signal as lircd.conf
                                                    # does not contain last off signal length
            if (i%6) ==0:
                fOut.write('\t\t\t\t')

            fOut.write('%d ' % round(self.dPulseWidths[i]))

            if (i+1)%6 ==0:      # Group codes as six per line
                fOut.write('\n')

        fOut.write('\n')          # Final EOL

class Device:   # Handles devices

    def AddCodes(self,sCodeName,sHexCodes):  # Add new code sequence
        seq=CodeSequence()
        finalgap=seq.AnalyzeCode(sCodeName,sHexCodes)

        if finalgap>self.lGap:
            self.lGap=finalgap

        self.sCodes.append(seq)

    def ProcessHEX(self,fHexFile,sLine):    # Process HEX files
        while sLine<>'' and sLine.strip()<>'':   # EOF?
            [sCodeName,sHexCodes]=sLine.split(':')
            self.AddCodes(sCodeName,sHexCodes)
            sLine=fHexFile.readline()

    def WriteLIRCCConfDevice(self,f):
        f.write('begin remote\n')
        f.write('\tname\t'+self.sDeviceName.replace(' ','')+'\n')
        f.write('\tflags\tRAW_CODES\n')
        f.write('\teps\t30\n')
        f.write('\taeps\t100\n')
        f.write('\tgap\t%d\n' % self.lGap )
        f.write('\t\tbegin raw_codes\n')

        for i in self.sCodes:
            i.WriteCodeSection(f)

        f.write('\t\tend raw_codes\n')
        f.write('end remote\n')

    def __init__(self,sDeviceName):
        self.sDeviceName=sDeviceName    # Name of the device
        self.sCodes=[]                  # Codes contained in file
        self.lGap=0                     # Final Gap

class RemoteFilesParser:
    def ProcessXML(self,fXMLFile):
        def start_element(name,attrs):
            if name=="RAWCODE":
                self.Devices[-1].AddCodes(attrs['name'],attrs['data'])
            if name=="DEVICE":
                self.Devices.append(Device(attrs['name']))

        p = xml.parsers.expat.ParserCreate()
        p.StartElementHandler = start_element
        fXMLFile.seek(0)        # Need to start from the beginning
        p.ParseFile(fXMLFile)

    def __init__(self,sFileName):
        self.Devices=[]

        f=open(sFileName,'r')
        sLine=f.readline()

        if sLine.strip()=='<?xml version="1.0"?>':   # Are we dealing with CCF Decompiler XML file?
                self.ProcessXML(f)
        else:
                device=Device(sFileName.split('.')[:1][0])
                self.Devices.append(device)
                device.ProcessHEX(f,sLine)

        f.close()

    def WriteLIRCConf(self,sOutFileName):
        f=open(sOutFileName,'w')

        for d in self.Devices:
            d.WriteLIRCCConfDevice(f)

        f.close()
# Main

import sys
import xml.parsers.expat

if len(sys.argv)<>2:
        print "Pronto codes converter to lircd.conf format (version 1.11)"
        print
        print "Usage:   pronto2lirc.py inputfile"
        print
        print "Inputfile can be:"
        print "         1.An xml file output by CCFTools/CCFDecompiler."
        print "         2.Text file where each line contains the name of the button and "
        print "         all codes associated with it"
        print "         Button1:0000 00ac 000b 00de ..."
        print
        print "Result:  lircd.conf file is written to the current directory"
        print "         containing all the Pronto codes extracted from"
        print "         the input file"
        print
else:
        p=RemoteFilesParser(sys.argv[1])
        p.WriteLIRCConf('lircd.conf')
